File: packages/expo-updates-interface/ios/EXUpdatesInterface/UpdatesInterface.swift
//  Copyright © 2021 650 Industries. All rights reserved.

import Foundation
import ExpoModulesCore

public typealias UpdatesErrorBlock = (_ error: Error) -> Void
public typealias UpdatesUpdateSuccessBlock = (_ manifest: [String: Any]?) -> Void
public typealias UpdatesProgressBlock = (_ successfulAssetCount: UInt, _ failedAssetCount: UInt, _ totalAssetCount: UInt) -> Void

/**
 * Called when a manifest has been downloaded. The return value indicates whether or not to
 * continue downloading the update described by this manifest. Returning `NO` will abort the
 * load, and the success block will be immediately called with a nil `manifest`.
 */
public typealias UpdatesManifestBlock = (_ manifest: [String: Any]) -> Bool

/**
 * Protocol for modules that depend on expo-updates for loading production updates but do not want
 * to depend on expo-updates or delegate control to the singleton EXUpdatesAppController.
 *
 * All updates controllers implement this protocol
 */
@objc(EXUpdatesInterface)
public protocol UpdatesInterface {
  @objc var runtimeVersion: String? { get }
  @objc var updateURL: URL? { get }
  @objc var isEnabled: Bool { get }
}

/**
 * Implemented only by the enabled updates controller
 */
public protocol UpdatesEnabledInterface: UpdatesInterface {
  var launchedUpdateId: UUID? { get }
  var embeddedUpdateId: UUID? { get }
  var stateChangeListener: (any UpdatesStateChangeListener)? { get set }
}

/**
 *
 * Implemented only by the dev launcher updates controller.
 */
@objc(EXUpdatesDevLauncherInterface)
public protocol UpdatesDevLauncherInterface: UpdatesInterface {
  @objc weak var updatesExternalInterfaceDelegate: (any UpdatesExternalInterfaceDelegate)? { get set }
  @objc var launchAssetURL: URL? { get }

  @objc var runtimeVersion: String? { get }
  @objc var updateURL: URL? { get }

  @objc func reset()

  @objc func fetchUpdate(
    withConfiguration configuration: [String: Any],
    onManifest manifestBlock: @escaping UpdatesManifestBlock,
    progress progressBlock: @escaping UpdatesProgressBlock,
    success successBlock: @escaping UpdatesUpdateSuccessBlock,
    error errorBlock: @escaping UpdatesErrorBlock
  )

  @objc func isValidUpdatesConfiguration(_ configuration: [String: Any]) -> Bool
}

/**
 * Protocol for communication/delegation back to the host dev client for functionality.
 */
@objc(EXUpdatesExternalInterfaceDelegate)
public protocol UpdatesExternalInterfaceDelegate {
  @objc func updatesExternalInterfaceDidRequestRelaunch(_ updatesExternalInterface: UpdatesDevLauncherInterface)
}

/**
 * Subclasses of this class can listen to state change events
 */
public protocol UpdatesStateChangeListener {
  func updatesStateDidChange(_ event: UpdatesStateEvent)
}

File: packages/expo-updates-interface/android/src/main/java/expo/modules/updatesinterface/UpdatesInterface.kt
package expo.modules.updatesinterface

import android.net.Uri
import expo.modules.updatesinterface.statemachine.UpdatesStateEvent
import org.json.JSONObject
import java.lang.ref.WeakReference
import java.util.UUID

/**
 * Interface for modules that depend on expo-updates for loading production updates but do not want
 * to depend on expo-updates or delegate control to the singleton UpdatesController.
 *
 * All updates controllers implement this protocol
 */
interface UpdatesInterface {
  val runtimeVersion: String?
  val updateUrl: Uri?
  val isEnabled: Boolean get() = false
}

/**
 * Implemented only by the enabled updates controller
 */
interface UpdatesEnabledInterface: UpdatesInterface {
  val launchedUpdateId: UUID? get() = null
  val embeddedUpdateId: UUID? get() = null
  var stateChangeListener: UpdatesStateChangeListener?
}

/**
 * Implemented only by the dev launcher updates controller.
 */
interface UpdatesDevLauncherInterface: UpdatesInterface {
  interface UpdateCallback {
    fun onFailure(e: Exception?)
    fun onSuccess(update: Update?)
    fun onProgress(successfulAssetCount: Int, failedAssetCount: Int, totalAssetCount: Int)

    /**
     * Called when a manifest has been downloaded. The return value indicates whether or not to
     * continue downloading the update described by this manifest. Returning `false` will abort the
     * load, and the `onSuccess` callback will be immediately called with a null `update`.
     */
    fun onManifestLoaded(manifest: JSONObject): Boolean
  }

  interface Update {
    val manifest: JSONObject
    val launchAssetPath: String
  }

  var updatesInterfaceCallbacks: WeakReference<UpdatesInterfaceCallbacks>?

  fun reset()
  fun fetchUpdateWithConfiguration(configuration: HashMap<String, Any>, callback: UpdateCallback)
  fun isValidUpdatesConfiguration(configuration: HashMap<String, Any>): Boolean
}

interface UpdatesStateChangeListener {
  fun updatesStateDidChange(event: UpdatesStateEvent)
}